src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let copyState = { ...state };
+  const arrayActions = [];
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      copyState = Object.assign({}, copyState, action.extraData);
+      arrayActions.push(copyState);
+    }
+
+    if (action.type === 'removeProperties') {
+      copyState = { ...copyState };
+
+      for (const property of action.keysToRemove) {
+        delete copyState[property];
+      }
+      arrayActions.push(copyState);
+    }
+
+    if (action.type === 'clear') {
+      copyState = {};
```

Comment:
nope, don't create new object, we will lose link to the previous object
It's better to use delete for each key in the object